File: app/src/main/java/com/example/olesya/boardgames/connection/CardHandler.java
package com.example.olesya.boardgames.connection;

import com.example.olesya.boardgames.Utils;

import java.io.IOException;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.Scanner;

public class CardHandler implements Runnable {
    private Server server;
    private PrintWriter outMessage;
    private Scanner inMessage;
    private String clientName;
    private int currentChoice = -1;
    private int totalPts = 0;

    public CardHandler(Socket socket, Server server) {
        try {
            this.server = server;
            this.outMessage = new PrintWriter(socket.getOutputStream());
            this.inMessage = new Scanner(socket.getInputStream());
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public void run() {
        try {
            acceptUserEvent();
            server.getScreenCallbacks().initHand(clientName);
            while (true) {
                if (inMessage.hasNext()) {
                    String clientMessage = inMessage.nextLine();
                    handleClientMessage(clientMessage);
                    if (clientMessage.equalsIgnoreCase(Utils.CLIENT_CONFIG.END_MSG)) {
                        break;
                    }
                }
                Thread.sleep(100);
            }
        } catch (InterruptedException ex) {
            ex.printStackTrace();
        } finally {
//            this.close();
        }
    }

    private void acceptUserEvent() {
        String username = inMessage.nextLine();
        this.clientName = server.getAvailableUsername(username, 0);
        if (!clientName.equals(username)) {
            sendMsg(Utils.CLIENT_CONFIG.USERNAME_CHANGED + Utils.DELIM + clientName);
        }

        server.getScreenCallbacks().onAddUserEvent(clientName);
    }

    private void handleClientMessage(String clientMessage) {
        String action = clientMessage.split("#")[0];
        String card;
        switch (action) {
            case Utils.CLIENT_COMMANDS.CLIENT_MAIN_FINISHED:
                card = clientMessage.split(Utils.DELIM)[1];
//                server.getCallbacks().onSelectedCardEvent(new Card(card, false, clientName));
                server.allowUsersToChoose(clientName);
//                server.showMessageFromClient(clientName, action);
                break;
            case Utils.CLIENT_COMMANDS.CLIENT_USER_FINISHED:
                card = clientMessage.split(Utils.DELIM)[1];
                break;
            case Utils.CLIENT_COMMANDS.CLIENT_USER_CHOOSE_FINISHED:
                currentChoice = Integer.valueOf(clientMessage.split(Utils.DELIM)[1]) - 1;
                server.getCallbacks().onAddUserChoice(clientName, currentChoice);
                server.checkForAllUsersChoice();
                break;
            case Utils.CLIENT_COMMANDS.CLIENT_MAIN_STOP_FINISHED:
                server.getScreenCallbacks().stopRound();
                break;
            case Utils.CLIENT_CONFIG.END_MSG:
                server.getScreenCallbacks().onRemoveUserEvent(clientName);
        }
    }

    public void sendMsg(String msg) {
        try {
            outMessage.println(msg);
            outMessage.flush();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    public void addPts(int pts) {
        totalPts += pts;
    }

    public String getName() {
        return clientName;
    }

    public int getChoice() {
        return currentChoice;
    }

    public int getPts() {
        return totalPts;
    }

    public void resetChoice() {
        currentChoice = -1;
    }
}

File: app/src/main/java/com/example/olesya/boardgames/connection/server/CardHandler.kt
package com.example.olesya.boardgames.connection.server

import android.util.Log
import com.example.olesya.boardgames.Commands
import com.example.olesya.boardgames.Commands.DELIM
import com.example.olesya.boardgames.game.controller.GameController
import com.example.olesya.boardgames.entity.Player
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.Socket

class CardHandler(socket: Socket, private var gameController: GameController) : Runnable {

    private val outMessage = PrintWriter(socket.getOutputStream())
    private val inMessage = BufferedReader(InputStreamReader(socket.getInputStream()))
    var username: String = ""

    override fun run() {
        acceptUserEvent()
        var clientMessage: String? = inMessage.readLine()
        while (clientMessage != null) {
            handleClientMessage(clientMessage)
            if (clientMessage.toUpperCase() == Commands.CLIENT_CONFIG.END_MSG) {
                gameController.onRemovePlayer(username)
            }

            clientMessage = inMessage.readLine()
        }
    }

    /**
     * Accept user and check for username
     */
    private fun acceptUserEvent() {
        username = inMessage.readLine()
        gameController.onAddUserEvent(username)
    }

    private fun handleClientMessage(clientMessage: String) {
        Log.d("Server", clientMessage)
        val action = clientMessage.split(DELIM)[0]
        gameController.screenMessage.postValue(action)
        when (action) {
            Commands.CLIENT_COMMANDS.CLIENT_TURN ->
                gameController.clientPicksCard(username, clientMessage.split(DELIM)[1])
            Commands.CLIENT_COMMANDS.CLIENT_CHOOSE ->
                gameController.clientChoosesCard(username, clientMessage.split(DELIM)[1])
        }
    }

    fun sendMsg(msg: String) {
        Thread {
            try {
                outMessage.println(msg)
                outMessage.flush()
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }.start()
    }
}